pkg/urbit/jets/e/en_json.c

```diff
@@ -0,0 +1,417 @@
+/* j/5/en_json.c
+**
+*/
+#include "all.h"
+
+/* helper macros
+*/
+
+  #ifndef SERIALIZE_BUFFER_INIT_SIZE
+    #define SERIALIZE_BUFFER_INIT_SIZE 4096
+  #endif
+
+  #ifndef SERIALIZE_BUFFER_INCR_SIZE
+    #define SERIALIZE_BUFFER_INCR_SIZE 4096
+  #endif
+
+/* helper structs
+*/
+
+  struct _json_buffer {
+    c3_y *buf_y;
+    c3_w len_w;
+    c3_w siz_w;
+  };
+
+/* constants
+*/
+
+  const c3_y _JSON_TRUE[] = "true";
+  const c3_y _JSON_FALSE[] = "false";
+  const c3_y _JSON_NULL[] = "null";
+  const c3_y _JSON_NEWLINE[] = "\\n";
+  const c3_y _JSON_DOQ[] = "\\\"";
+  const c3_y _JSON_BAS[] = "\\\\";
+  const c3_y *_JSON_UNICODES[] = {
+    "\\u0000",  // 0
+    "\\u0001",  // 1
+    "\\u0002",  // 2
+    "\\u0003",  // 3
+    "\\u0004",  // 4
+    "\\u0005",  // 5
+    "\\u0006",  // 6
+    "\\u0007",  // 7
+    "\\u0008",  // 8
+    "\\u0009",  // 9
+    "\\u000a",  // 10
+    "\\u000b",  // 11
+    "\\u000c",  // 12
+    "\\u000d",  // 13
+    "\\u000e",  // 14
+    "\\u000f",  // 15
+    "\\u0010",  // 16
+    "\\u0011",  // 17
+    "\\u0012",  // 18
+    "\\u0013",  // 19
+    "\\u0014",  // 20
+    "\\u0015",  // 21
+    "\\u0016",  // 22
+    "\\u0017",  // 23
+    "\\u0018",  // 24
+    "\\u0019",  // 25
+    "\\u001a",  // 26
+    "\\u001b",  // 27
+    "\\u001c",  // 28
+    "\\u001d",  // 29
+    "\\u001e",  // 30
+    "\\u001f",  // 31
+  };
+  const c3_y _JSON_DEL[] = "\\u007f";
+
+/* forward declarations
+*/
+
+  struct _json_buffer*
+  _json_serialize(struct _json_buffer*, u3_noun);
+
+/* core jet logic
+*/
+
+  struct _json_buffer*
+  _json_en_init()
+  {
+    struct _json_buffer *buf_u = (struct _json_buffer*)u3a_malloc(sizeof(struct _json_buffer));
+    if (buf_u == NULL)
+      return NULL;
+
+    buf_u->buf_y = (c3_y*)u3a_malloc(SERIALIZE_BUFFER_INIT_SIZE);
+    buf_u->len_w = 0;
+    buf_u->siz_w = SERIALIZE_BUFFER_INIT_SIZE;
+
+    if (buf_u->buf_y == NULL) {
+      u3a_free(buf_u);
+      return NULL;
+    }
+    return buf_u;
+  }
+
+  void
+  _json_en_close(struct _json_buffer *buf_u)
+  {
+    if (buf_u->buf_y != NULL)
+      u3a_free(buf_u->buf_y);
+    u3a_free(buf_u);
+  }
+
+  struct _json_buffer*
+  _json_en_fail(struct _json_buffer *buf_u)
+  {
+    _json_en_close(buf_u);
+    return NULL;
+  }
+
+  struct _json_buffer*
+  _json_resize(struct _json_buffer *buf_u)
+  {
+    if (buf_u == NULL)
+      return NULL;
+
+    buf_u->siz_w += SERIALIZE_BUFFER_INCR_SIZE;
+    buf_u->buf_y = (c3_y*)u3a_realloc(buf_u->buf_y, buf_u->siz_w);
+    if (buf_u->buf_y == NULL)
+      return _json_en_fail(buf_u);
+    return buf_u;
+  }
```

Comment:
> Separate functions for measure and cut could be more efficient, but that might be too much

"Too much" in what sense? This bounty is a means for me to learn about the runtime, so if separate functions are more efficient and also have correct allocation (for jets), then I'd prefer to do it that way even if it's challenging

> I think `+en-json` escapes more than is strictly necessary

Could you elaborate? Do you mean changing it to only escape `\n`, `"`, and `\` , and writing everything else as UTF8?

Reply:
"Too much" in the overoptimization sense. I think the right way to proceed is to first polish up what you have, and then experiment with some further optimization of allocations. Could profile the jet as is, try `mcut`-style, and then maybe try separating out measurement and serialization functions (depending on the profile results).

Calculating the escape size is going to require a full scan regardless (which should be pretty fast).

Reply:
I measured the JSON encoding performance on a large JSON object (Twitter user data) which I've added to the test cases. Here are the results:
```
TEST ENVIRONMENT                  MIN   AVG   MAX
---------------------------------------------------------
Fresh dev planet, latest binary:  3610  3920  4445
Current commit, jet disabled:     2108  2210  3046
Current commit, jet enabled:      423   529   881
Measure, parse, jet, heap:        435   553   1260
Measure, parse, jet, stack:       448   547   863
```
These numbers are the wall-clock times in `µs` from 20 runs of the large encoding unit test.

Even with a tweets worth of data, the final size of the byte array was ~2800, so it didn't hit a resize call. For that reason, I can't imagine `mcut` improving performance over "current commit, jet enabled". In addition, because of all the `if` and `switch` statements, I think `mcut` functions will end up looking pretty ugly.

I shared this info with @joemfb and we agreed that we should go with separate measure and parse functions, though there will be a little repetition.